Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,564 @@
+    interceptors: [ServerInterceptor<Request, Response>],
+    userHandler: @escaping @Sendable(
+      GRPCAsyncRequestStream<Request>,
+      GRPCAsyncResponseStreamWriter<Response>,
+      GRPCAsyncServerCallContext
+    ) async throws -> Void
+  ) {
+    self.serializer = responseSerializer
+    self.deserializer = requestDeserializer
+    self.context = context
+    self.userHandler = userHandler
+
+    let userInfoRef = Ref(UserInfo())
+    self.userInfoRef = userInfoRef
+
+    self.interceptors = ServerInterceptorPipeline(
+      logger: context.logger,
+      eventLoop: context.eventLoop,
+      path: context.path,
+      callType: .bidirectionalStreaming,
+      remoteAddress: context.remoteAddress,
+      userInfoRef: userInfoRef,
+      interceptors: interceptors,
+      onRequestPart: self.receiveInterceptedPart(_:),
+      onResponsePart: self.sendInterceptedPart(_:promise:)
+    )
+  }
+
+  // MARK: - GRPCServerHandlerProtocol conformance
+
+  @inlinable
+  internal func receiveMetadata(_ headers: HPACKHeaders) {
+    self.interceptors.receive(.metadata(headers))
+  }
+
+  @inlinable
+  internal func receiveMessage(_ bytes: ByteBuffer) {
+    do {
+      let message = try self.deserializer.deserialize(byteBuffer: bytes)
+      self.interceptors.receive(.message(message))
+    } catch {
+      self.handleError(error)
+    }
+  }
+
+  @inlinable
+  internal func receiveEnd() {
+    self.interceptors.receive(.end)
+  }
+
+  @inlinable
+  internal func receiveError(_ error: Error) {
+    self.handleError(error)
+    self.finish()
+  }
+
+  @inlinable
+  internal func finish() {
+    switch self.state {
+    case .idle:
+      self.interceptors = nil
+      self.state = .completed
+
+    case let .active(_, _, statusPromise):
+      statusPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+
+    case .completed:
+      self.interceptors = nil
+    }
+  }
+
+  // MARK: - Interceptors to User Function
+
+  @inlinable
+  internal func receiveInterceptedPart(_ part: GRPCServerRequestPart<Request>) {
+    switch part {
+    case let .metadata(headers):
+      self.receiveInterceptedMetadata(headers)
+    case let .message(message):
+      self.receiveInterceptedMessage(message)
+    case .end:
+      self.receiveInterceptedEnd()
+    }
+  }
+
+  @inlinable
+  internal func receiveInterceptedMetadata(_ headers: HPACKHeaders) {
+    switch self.state {
+    case .idle:
+      let statusPromise: EventLoopPromise<GRPCStatus> = context.eventLoop.makePromise()
+
+      // Make a context to invoke the user handler with.
+      let context = GRPCAsyncServerCallContext(
+        headers: headers,
+        logger: self.context.logger,
+        userInfoRef: self.userInfoRef
+      )
+
+      // Create a request stream to pass to the user function and capture the
+      // handler in the updated state to allow us to produce more results.
+      let requestStream = GRPCAsyncRequestStream<Request>(AsyncThrowingStream { continuation in
```

Comment:
FYI https://github.com/grpc/grpc-swift/pull/1252 has been merged now. It has two parts a `PassthroughMessageSource` and a `PassthroughMessageSequence` (an adapter for the source to turn it into an `AsyncSequence`)

Reply:
I'm don't mind. It shouldn't be a particularly large change though.